[leehuun][react-week2-assignment-1]

huun0424 opened the pull request:
안녕하세요 :)
죄송합니다. 선약때문에 바빠서 지금에서야 PR 올립니다.
2번 과제는 내일 작업할건데, 늦은 밤에나 시작할 수 있을거같아요 ㅠㅜ.

작업내용.
 
- Button component 추가
 - eslint 설정
 - index에 연산 로직 추가

huun0424 committed:
[leehuun][react-week2-assignment-1]
 - Button component 추가
 - eslint 설정
 - index에 연산 로직 추가

passwd10 commented on src/index.jsx:
```diff
@@ -0,0 +1,32 @@
+import ReactDom from 'react-dom';
+import Button from './Button';
+
+const INCREMENT_LIST = [1, 2, 3, 4, 5];
```
LIST라는 표현대신 좀 더 의미가 명확한 표현을 사용해보세요.

passwd10 commented on src/Button.jsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
```
react 17 버전부터는 생략이 가능해요.

현재 사용중인 라이브러리의 버전을 확인하고 추가된 기능을 확인해보세요 :)

[whats-different-in-the-new-transform](https://ko.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#whats-different-in-the-new-transform)

passwd10 commented on src/Button.jsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+const Button = ({ children, onClick }) => (
```
children 은 보통 ReactNode를 넘길 때 사용하는 표현이에요. button의 value를 표현할 수 있는 단어를 사용해보세요.

huun0424 replied:
숫자나 문자열 외에도 ReactNode도 들어올 수도 있겠다 싶어서 사용 했었어요. 피드백 주신대로 ReactNode가 들어오지 않으니 수정 해놓겠습니다 :)

passwd10 commented on src/index.jsx:
```diff
@@ -0,0 +1,32 @@
+const App = () => {
+  const [count, setCount] = useState(0);
+
+  const handleClick = useCallback((increment) => {
+    setCount((prev) => (prev + increment));
+  }, [count]);
```
useCallback을 사용해 memoization 하셨네요 👍

passwd10 commented on src/index.jsx:
```diff
@@ -0,0 +1,32 @@
+        {`Click me (${count})`}
+      </Button>
+      <div>
+        {
+          INCREMENT_LIST.map((increment) => (
+            <Button key={`increment-button-${increment}`} onClick={() => handleClick(increment)}>
+              { increment }
+            </Button>
+          ))
+        }
```
`Button` 컴포넌트를 감싸고 있는 리스트도 컴포넌트로 분리할 수 있을 것 같아요

passwd10 commented on src/index.jsx:
```diff
@@ -0,0 +1,32 @@
+import React, { useCallback, useState } from 'react';
+import ReactDom from 'react-dom';
+import Button from './Button';
```
import 하는 모듈들의 의미와 역할이 다르기 때문에 공백으로 구분해주시면 가독성이 좋아질 것 같아요.

passwd10 commented on src/index.jsx:
```diff
@@ -0,0 +1,32 @@
+
+const INCREMENT_LIST = [1, 2, 3, 4, 5];
+
+const App = () => {
```
App 컴포넌트도 하나의 파일로 분리해보세요.

passwd10 reviewed (commented):
바쁘신 와중에도 과제를 잘 제출해주셨네요 😄 
지금은 Button 컴포넌트만 나누셨는데, 다른 컴포넌트들도 분리해보세요.
분리하시면서 왜 분리해야하는지? 어떤 의미가 있는지를 고민하시면 좋을 것 같아요!

passwd10 commented on src/Button.jsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+const Button = ({ children, onClick }) => (
+  <button type="button" onClick={onClick || null}>
```
null 은 어떤 의도로 넣으신건가요~?

huun0424 replied:
예외 처리를 하려고 넣었는데, 보니까 필요없는거 같아요. 수정하겠습니다. 감사합니다 :)

passwd10 commented on src/index.jsx:
```diff
@@ -0,0 +1,32 @@
+const App = () => {
+  const [count, setCount] = useState(0);
+
+  const handleClick = useCallback((increment) => {
+    setCount((prev) => (prev + increment));
+  }, [count]);
```
useCallback을 사용하신 이유를 알 수 있을까요? 의존성 배열에 count 를 넣어주셨는데 어떤 현상이 발생할까요?

huun0424 replied:
React가 render 될때 마다 내부 함수들이 재선언 된다고 알고있어서, 불필요한 재선언을 막기위해 useCallback을 사용했어요. 혹시 이벤트는 useCallback으로 최적화하는 과정이 필요 없나요? count는 setCount를 넣어야하는데 잘못 들어갔네요.

passwd10 replied:
이벤트 핸들러가 최적화의 조건이 된다기보단, 현 상황에서 최적화와 관련된 trade-off 를 고민해보시면 좋을 것 같아요. useCallback을 사용했을 때 무엇을 얻을 수 있고, 어떤 비용이 들까요? 고민해보시고 어떻게 검증할 수 있을지 생각해보시면 좋을 것 같아요!

추가로 의존성 배열에 setState 함수를 넣어주지 않아도 됩니다! 아래 문서를 읽어보시면 도움이 될 것 같아요~

- https://reactjs.org/docs/hooks-reference.html#basic-hooks

Commit:
[leehuun][react-week2-assignment-1]
 - App, IncrementButtons로 comp 분리
 - React babel 설정 추가
 - const 변수 리네임
 - useCallback 제거

huun0424 commented:
안녕하세요. :)

피드백 주신내용들 학습 후, 리팩토린 했습니다.

passwd10 commented on src/App.jsx:
```diff
@@ -0,0 +1,24 @@
+import Button from './Button';
+import IncrementButtons from './IncrementButtons';
+
+const INCREMENT_NUMBER_LIST = [1, 2, 3, 4, 5];
```
숫자값들을 상수로 분리하신 아이디어 좋네요 👍 

만약 숫자들을 유동적으로 변경해야하는 상황이 온다면 어떻게 대응해야 할까요?

huun0424 replied:
동적으로 변경해야하는 상황이 온다면 state로 상태관리를 하면 되지 않을까요? :)
